packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx

```diff
@@ -0,0 +1,156 @@
+    (params, event) => {
+      if (isRowReorderDisabled) {
+        return;
+      }
+
+      logger.debug(`Start dragging row ${params.id}`);
+      // Prevent drag events propagation.
+      // For more information check here https://github.com/mui/mui-x/issues/2680.
+      event.stopPropagation();
+
+      dragRowNode.current = event.currentTarget;
+      dragRowNode.current.classList.add(classes.rowDragging);
+
+      setDragRowId(params.id);
+
+      removeDnDStylesTimeout.current = setTimeout(() => {
+        dragRowNode.current!.classList.remove(classes.rowDragging);
+      });
+
+      originRowIndex.current = apiRef.current.getRowIndex(params.id);
+    },
+    [isRowReorderDisabled, classes.rowDragging, logger, apiRef],
+  );
+
+  const handleDragOver = React.useCallback<
+    GridEventListener<GridEvents.cellDragOver | GridEvents.rowDragOver>
+  >(
+    (params, event) => {
+      if (!dragRowId && dragRowId !== 0) {
+        return;
+      }
+      logger.debug(`Dragging over row ${params.id}`);
+      event.preventDefault();
+      // Prevent drag events propagation.
+      // For more information check here https://github.com/mui/mui-x/issues/2680.
+      event.stopPropagation();
+
+      if (params.id !== dragRowId) {
+        const targetRowIndex = apiRef.current.getRowIndex(params.id);
+        const dragRowIndex = apiRef.current.getRowIndex(dragRowId);
+
+        apiRef.current.setRowIndex(dragRowId, targetRowIndex);
+
+        const rowOrderChangeParams: GridRowOrderChangeParams = {
+          row: apiRef.current.getRow(dragRowId),
+          targetIndex: targetRowIndex,
+          oldIndex: dragRowIndex,
+        };
+        apiRef.current.publishEvent(GridEvents.rowOrderChange, rowOrderChangeParams);
+      }
+    },
+    [apiRef, logger, dragRowId],
+  );
+
+  const handleDragEnd = React.useCallback<GridEventListener<GridEvents.rowDragEnd>>(
+    (params, event): void => {
+      if (isRowReorderDisabled || (!dragRowId && dragRowId !== 0)) {
+        return;
+      }
+
+      logger.debug('End dragging row');
+      event.preventDefault();
+      // Prevent drag events propagation.
+      // For more information check here https://github.com/mui/mui-x/issues/2680.
+      event.stopPropagation();
+
+      clearTimeout(removeDnDStylesTimeout.current);
+      dragRowNode.current = null;
+
+      // Check if the row was dropped outside the grid.
+      if (event.dataTransfer.dropEffect === 'none') {
+        // Accessing params.field may contain the wrong field as header elements are reused
+        apiRef.current.setRowIndex(dragRowId, originRowIndex.current!);
+
+        const rowOrderChangeParams: GridRowOrderChangeParams = {
+          row: apiRef.current.getRow(dragRowId),
+          targetIndex: originRowIndex.current!,
+          oldIndex: originRowIndex.current!,
+        };
+        apiRef.current.publishEvent(GridEvents.rowOrderChange, rowOrderChangeParams);
+
+        originRowIndex.current = null;
+      }
```

Comment:
I updated it so that the rowOrderChange is only emitted when the dragging stops. I think that would be optimal as you don't want to get spammed with multiple reorders, we should only care about the last position.